Add NodeStage error tests

jsafrane opened the pull request:
**What this PR does / why we need it**:
Add some test for `NodeStage` error handling. The main purpose is to test that:
* `NodeUnstage` **is** called after `NodeStage` transient error && corresponding pod is deleted.
* `NodeUnstage` **is not** called after `NodeStage` final error && corresponding pod is deleted.

The test (and whole CSI mock output handling) becomes quite complex.

This is just an exercise how to use the javascript hooks. If we decide this is useful, it can be extended to test also NodePublish (+ block mode of both).

@gnufied @tsmetana @pohly @msau42 @jingxu97, *is* it useful? We already have unit tests for this behavior.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/kind cleanup
/sig storage

jsafrane commented:
the newly introduced tests are flaky, https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89041/pull-kubernetes-e2e-gce-storage-slow/1237750118531207170/:

```
test/e2e/storage/csi_mock_volume.go:664
Mar 11 15:26:33.230: while waiting for initial CSI calls
Unexpected error:
    <*errors.errorString | 0xc001da7a10>: {
        s: "could not load CSI driver logs: the server rejected our request for an unknown reason (get pods csi-mockplugin-0)",
    }
    could not load CSI driver logs: the server rejected our request for an unknown reason (get pods csi-mockplugin-0)
occurred
test/e2e/storage/csi_mock_volume.go:688
```
/hold
/retest

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
Reworked to merged version of javascript hooks.

The only WIP item is bump of csi-mock driver version to `canary`. We should do a proper release before merging this PR.

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow
/retest

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
/hold cancel
Rebased to mock v3.1.0 with the javascript hooks.

pohly commented:
Is the new test flaky? There's a "could not load CSI driver logs: the server rejected our request for an unknown reason (get pods csi-mockplugin-0)" in https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89041/pull-kubernetes-e2e-gce-storage-slow/1245284273221537792/.

/retest

pohly commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -768,6 +935,84 @@ func checkPodLogs(cs clientset.Interface, namespace, driverPodName, driverContai
+	// Load logs of driver pod
+	log, err := e2epod.GetPodLogs(cs, namespace, driverPodName, driverContainerName)
+	if err != nil {
+		return nil, fmt.Errorf("could not load CSI driver logs: %s", err)
```
This failed because "the server rejected our request for an unknown reason (get pods csi-mockplugin-0)" (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89041/pull-kubernetes-e2e-gce-storage-slow/1245284273221537792/).

Perhaps a retry loop would help?

pohly replied:
There is already a retry loop in the caller (via `wait.Poll`). Perhaps failures to retrieve log output should simply be treated here as "no output", i.e. return `nil, nil`?

jsafrane replied:
Maybe the pod is not Running yet. The new mock test calls `createPod()` and immediately after that it reads the pod logs. I added wait for PVC to get bound in between - the driver must be fully operational to provision a PV.

jsafrane replied:
Sorry about the failed build noise. Stupid typo...

jsafrane replied:
I don't get it. The mock driver provisioned a PV, yet it gets error from API server:

`
could not load CSI driver logs: the server rejected our request for an unknown reason (get pods csi-mockplugin-0)
`

pohly replied:
> There is already a retry loop in the caller (via wait.Poll). Perhaps failures to retrieve log output should simply be treated here as "no output", i.e. return nil, nil?

I've implemented that idea in https://github.com/kubernetes/kubernetes/pull/88114 (https://github.com/kubernetes/kubernetes/pull/88114/commits/3cf9ab134a3201880472c063e5b527b88f3b0ebb) and got all tests to pass. The log does indeed show that "the server rejected our request for an unknown reason" occurred, but tests succeeded after ignoring that error (https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88114/pull-kubernetes-e2e-kind/1245706940810530817/build-log.txt).

jsafrane replied:
I added your last commit from #88114, hoping it helps.

I can't reproduce "the server rejected our request for an unknown reason" on any of my clusters.

pohly replied:
> I can't reproduce "the server rejected our request for an unknown reason" on any of my clusters.

Me neither. I think saw it once while working on some other test, but as far as I remember, that then turned out to be because I was I was asking for logs after the pod had just been deleted. Perhaps here we have the inverse, asking for a very recently started pod? Just wondering.

jsafrane commented:
It comes from https://github.com/kubernetes-csi/external-attacher/pull/155
Exact reason is not listed there. I don't see anything in the CHANGELOG that needs Kubernetes 1.15.

gnufied commented:
/assign

jsafrane committed:
Test NodeStage error cases

Especially related to "uncertain" global mounts. A large refactoring of CSI
mock tests were necessary:
- to be able to script the driver to return errors as required by the test
- to parse the CSI driver logs to check kubelet called the right CSI calls

jsafrane committed:
Update to new javascript

pohly committed:
mock tests: do not give up immediately for pod output errors

As seen in some test
runs (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89041),
retrieving output can fail with "the server rejected our request for
an unknown reason (get pods csi-mockplugin-0)".

If this truly an intermittent error, then the existing retry logic in
the callers can deal with this.

pohly commented:
/retest

jsafrane commented:
/retest

pohly reviewed (commented):
/lgtm
